File: 0002_两数相加.go
package leetcode

/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */

// 这种解法对链表很长的情况会因为整数溢出不通过
// func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
// 	n1 := 0
// 	n2 := 0
// 	for i := 0; l1 != nil; i++ {
// 		n1 += int(math.Pow10(i)) * l1.Val
// 		l1 = l1.Next
// 	}
// 	for i := 0; l2 != nil; i++ {
// 		n2 += int(math.Pow10(i)) * l2.Val
// 		l2 = l2.Next
// 	}
// 	sum := n1 + n2
// 	var head, tail *ListNode
// 	if sum == 0 {
// 		return &ListNode{Val: 0}
// 	}
// 	for sum > 0 {
// 		r := sum % 10
// 		ptr := &ListNode{Val: r}
// 		if head == nil {
// 			head = ptr
// 			tail = ptr
// 		} else {
// 			tail.Next = ptr
// 			tail = tail.Next
// 		}
// 		sum /= 10
// 	}
// 	// fmt.Println("n1", n1, "n2", n2)
// 	return head
// }

func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
	var head, tail *ListNode
	carry := 0 // 上一次是否有进位
	for l1 != nil && l2 != nil {
		p := &ListNode{Val: (l1.Val + l2.Val + carry) % 10}
		if head == nil {
			head = p
			tail = p
		} else {
			tail.Next = p
			tail = tail.Next
		}
		if l1.Val+l2.Val+carry >= 10 {
			carry = 1
		} else {
			carry = 0
		}
		l1 = l1.Next
		l2 = l2.Next
	}
	// 加上剩下的部分
	for l1 != nil {
		p := &ListNode{Val: (l1.Val + carry) % 10}
		tail.Next = p
		tail = tail.Next
		if l1.Val+carry >= 10 {
			carry = 1
		} else {
			carry = 0
		}
		l1 = l1.Next
	}
	for l2 != nil {
		p := &ListNode{Val: (l2.Val + carry) % 10}
		tail.Next = p
		tail = tail.Next
		if l2.Val+carry >= 10 {
			carry = 1
		} else {
			carry = 0
		}
		l2 = l2.Next
	}
	// 若最后有进位位, 添加上
	if carry == 1 {
		p := &ListNode{Val: 1}
		tail.Next = p
		tail = tail.Next
	}
	return head
}

File: python/0002_两数相加.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None


class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        v1 = []
        v2 = []
        while l1:
            v1.append(l1.val)
            l1 = l1.next
        while l2:
            v2.append(l2.val)
            l2 = l2.next
        v1.reverse()
        v2.reverse()
        s1 = 0
        s2 = 0
        for i in v1:
            s1 = s1 * 10 + i
        for j in v2:
            s2 = s2 * 10 + j
        sum = s1 + s2
        sl = list(str(sum))
        sl = list(map(int, sl))
        sl.reverse()
        head = ListNode(0)
        p = head

        for v in sl:
            tmp = ListNode(v)
            p.next = tmp
            p = p.next
        return head.next
